code/modules/reagents/Chemistry-Machinery.dm

```diff
@@ -85,247 +96,420 @@
-	if(stat & (BROKEN|NOPOWER)) return
-	if(usr.stat || usr.restrained()) return
-	if(!in_range(src, usr)) return
-
-	src.add_fingerprint(usr)
-	usr.set_machine(src)
-
-	if(href_list["tab_select"])
-		tab = href_list["tab_select"]
-
-	if (href_list["ejectp"])
-		if(loaded_pill_bottle)
-			loaded_pill_bottle.forceMove(get_turf(src))
+	// Transfer modal information if there is one
+	data["modal"] = tgui_modal_data(src)
 
-			if(Adjacent(usr))
-				usr.put_in_hands(loaded_pill_bottle)
+	return data
 
-			loaded_pill_bottle = null
-
-	if(beaker)
-		var/datum/reagents/R = beaker.reagents
-		if (tab == "analyze")
-			analyze_data["name"] = href_list["name"]
-			analyze_data["desc"] = href_list["desc"]
-			if(!condi)
-				if(href_list["name"] == "Blood")
-					var/datum/reagent/blood/G
-					for(var/datum/reagent/F in R.reagent_list)
-						if(F.name == href_list["name"])
-							G = F
-							break
-					analyze_data["name"] = G.name
-					analyze_data["blood_type"] = G.data["blood_type"]
-					analyze_data["blood_DNA"] = G.data["blood_DNA"]
-
-		else if (href_list["add"])
-
-			if(href_list["amount"])
-				var/id = href_list["add"]
-				var/amount = CLAMP((text2num(href_list["amount"])), 0, 200)
-				R.trans_id_to(src, id, amount)
-
-		else if (href_list["addcustom"])
-
-			var/id = href_list["addcustom"]
-			useramount = input("Select the amount to transfer.", 30, useramount) as num
-			useramount = CLAMP(useramount, 0, 200)
-			src.Topic(null, list("amount" = "[useramount]", "add" = "[id]"))
-
-		else if (href_list["remove"])
-
-			if(href_list["amount"])
-				var/id = href_list["remove"]
-				var/amount = CLAMP((text2num(href_list["amount"])), 0, 200)
-				if(mode)
-					reagents.trans_id_to(beaker, id, amount)
+/**
+  * Called in tgui_act() to process modal actions
+  *
+  * Arguments:
+  * * action - The action passed by tgui
+  * * params - The params passed by tgui
+  */
+/obj/machinery/chem_master/proc/tgui_act_modal(action, params, datum/tgui/ui, datum/tgui_state/state)
+	. = TRUE
+	var/id = params["id"] // The modal's ID
+	var/list/arguments = istext(params["arguments"]) ? json_decode(params["arguments"]) : params["arguments"]
+	switch(tgui_modal_act(src, action, params))
+		if(TGUI_MODAL_OPEN)
+			switch(id)
+				if("analyze")
+					var/idx = text2num(arguments["idx"]) || 0
+					var/from_beaker = text2num(arguments["beaker"]) || FALSE
+					var/reagent_list = from_beaker ? beaker.reagents.reagent_list : reagents.reagent_list
+					if(idx < 1 || idx > length(reagent_list))
+						return
+
+					var/datum/reagent/R = reagent_list[idx]
+					var/list/result = list("idx" = idx, "name" = R.name, "desc" = R.description)
+					if(!condi && istype(R, /datum/reagent/blood))
+						var/datum/reagent/blood/B = R
+						result["blood_type"] = B.data["blood_type"]
+						result["blood_dna"] = B.data["blood_DNA"]
+
+					arguments["analysis"] = result
+					tgui_modal_message(src, id, "", null, arguments)
+				// if("change_pill_bottle_style")

```

Comment:
Is this a missing function, or...?

Reply:
We don't have custom pill bottle sprites, Paradise (where this code was taken from) does. I figured I'd just leave it in in case someone or me goes back and adds them later.

Reply:
I see!